src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+        /// Number of threads used for decompression.
+        ///
+        /// Default value: 4.
+        ///
+        /// Note that multiple times this many threads will be spawned by [`split_extract()`] as
+        /// part of the pipelined process. Only this many threads will be used to perform
+        /// decompression in rust code, but other threads will be used to wait on I/O from
+        /// the kernel.
+        pub decompression_threads: usize,
+        /// Size of buffer used to copy a decompressed entry into the corresponding output pipe.
+        ///
+        /// Default value: 1MB.
+        pub decompression_copy_buffer_length: usize,
+        /// Size of buffer used to copy stored entries into the output file.
+        ///
+        /// Used on non-Linux platforms without
+        /// [`copy_file_range()`](https://www.gnu.org/software/libc/manual/html_node/Copying-File-Data.html),
+        /// as well as on Linux when the input and output file handles are on separate devices.
+        ///
+        /// Default value: 1MB.
+        pub file_range_copy_buffer_length: usize,
+        /// Size of buffer used to splice contents from a pipe into an output file handle.
+        ///
+        /// Used on non-Linux platforms without [`splice()`](https://en.wikipedia.org/wiki/Splice_(system_call)).
+        ///
+        /// Default value: 1MB.
+        #[cfg(not(target_os = "linux"))]
+        pub splice_read_buffer_length: usize,
+        /// Size of buffer used to splice contents from an input file handle into a pipe.
+        ///
+        /// Used on non-Linux platforms without [`splice()`](https://en.wikipedia.org/wiki/Splice_(system_call)).
+        ///
+        /// Default value: 1MB.
+        #[cfg(not(target_os = "linux"))]
+        pub splice_write_buffer_length: usize,
+    }
+
+    impl Default for ExtractionParameters {
+        fn default() -> Self {
+            Self {
+                decompression_threads: 4,
+                decompression_copy_buffer_length: 1024 * 1024,
+                file_range_copy_buffer_length: 1024 * 1024,
+                #[cfg(not(target_os = "linux"))]
+                splice_read_buffer_length: 1024 * 1024,
+                #[cfg(not(target_os = "linux"))]
+                splice_write_buffer_length: 1024 * 1024,
+            }
+        }
+    }
+
+    fn wrap_spawn_err<'scope>(
+        err_sender: mpsc::Sender<SplitExtractionError>,
+        f: impl FnOnce() -> Result<(), SplitExtractionError> + Send + 'scope,
+    ) -> impl FnOnce() + Send + 'scope {
+        move || match f() {
+            Ok(()) => (),
+            #[allow(clippy::single_match)]
+            Err(e) => match err_sender.send(e) {
+                Ok(()) => (),
+                /* We use an async sender, so this should only error if the receiver has hung
+                 * up, which occurs when we return a previous error from the main thread. */
+                Err(mpsc::SendError(_)) => (),
+            },
+        }
```

Comment:
```suggestion
        move || if let Err(e) = f() {
            /* We use an async sender, so this should only error if the receiver has hung up because of a previous error. */
                let _ = err_sender.send(e);
            }
        }
```

Reply:
Done!